File: codex_status.js
#!/usr/bin/env node

const pty = require("node-pty");

function parseArgs(argv) {
  const args = {
    codexPath: "codex",
    slowDelay: 0.1,
    readyTimeout: 30,
    statusTimeout: 30,
    debug: false,
  };

  for (let i = 2; i < argv.length; i += 1) {
    const a = argv[i];
    if (a === "--codex-path") {
      i += 1;
      if (i >= argv.length) throw new Error("--codex-path requires a value");
      args.codexPath = argv[i];
    } else if (a === "--slow-delay") {
      i += 1;
      if (i >= argv.length) throw new Error("--slow-delay requires a value");
      args.slowDelay = Number(argv[i]);
    } else if (a === "--ready-timeout") {
      i += 1;
      if (i >= argv.length) throw new Error("--ready-timeout requires a value");
      args.readyTimeout = Number(argv[i]);
    } else if (a === "--status-timeout") {
      i += 1;
      if (i >= argv.length) throw new Error("--status-timeout requires a value");
      args.statusTimeout = Number(argv[i]);
    } else if (a === "--debug") {
      args.debug = true;
    } else if (a === "-h" || a === "--help") {
      printHelp();
      process.exit(0);
    } else {
      throw new Error(`Unknown argument: ${a}`);
    }
  }

  return args;
}

function printHelp() {
  const text = [
    "Usage: node codex_status.js [options]",
    "",
    "Options:",
    "  --codex-path <path>      Path to codex executable (default: codex)",
    "  --slow-delay <seconds>   Delay per typed character (default: 0.10)",
    "  --ready-timeout <sec>    Wait for Codex UI ready (default: 30)",
    "  --status-timeout <sec>   Wait for /status output (default: 30)",
    "  --debug                  Enable debug logs to stderr",
    "  -h, --help               Show this help",
  ].join("\n");
  process.stdout.write(`${text}\n`);
}

function makeLogger(debug) {
  return (msg) => {
    if (!debug) return;
    const d = new Date();
    const offMin = -d.getTimezoneOffset();
    const sign = offMin >= 0 ? "+" : "-";
    const abs = Math.abs(offMin);
    const hh = String(Math.floor(abs / 60)).padStart(2, "0");
    const mm = String(abs % 60).padStart(2, "0");
    const ts = `${d.getFullYear()}-${String(d.getMonth() + 1).padStart(2, "0")}-${String(d.getDate()).padStart(2, "0")} ${String(d.getHours()).padStart(2, "0")}:${String(d.getMinutes()).padStart(2, "0")}:${String(d.getSeconds()).padStart(2, "0")}${sign}${hh}${mm}`;
    process.stderr.write(`[${ts}] ${msg}\n`);
  };
}

const ANSI_RE = /\x1b\[[0-9;]*[A-Za-z]/g;
const BOX_RE = /[│╭╮╰╯─•╳]/g;

function normalizeCapture(raw) {
  const deansi = raw.replace(ANSI_RE, "").replace(BOX_RE, " ");
  const lines = [];
  for (const ln of deansi.split(/\r?\n/)) {
    const s = ln.replace(/\s+/g, " ").trim();
    if (s) lines.push(s);
  }
  return lines;
}

const MONTHS = {
  Jan: 1,
  Feb: 2,
  Mar: 3,
  Apr: 4,
  May: 5,
  Jun: 6,
  Jul: 7,
  Aug: 8,
  Sep: 9,
  Oct: 10,
  Nov: 11,
  Dec: 12,
};

function nextLocalTimeToEpoch(hhmm) {
  const [h, m] = hhmm.split(":").map((v) => Number(v));
  const now = new Date();
  const cand = new Date(now);
  cand.setHours(h, m, 0, 0);
  if (cand <= now) cand.setDate(cand.getDate() + 1);
  return Math.floor(cand.getTime() / 1000);
}

function parseNoYearToEpoch(s) {
  const m = s.trim().match(/^(\d{1,2}):(\d{2})\s+on\s+(\d{1,2})\s+([A-Za-z]{3})$/);
  if (!m) throw new Error(`Unrecognized reset format: ${s}`);
  const hh = Number(m[1]);
  const mm = Number(m[2]);
  const dd = Number(m[3]);
  const mon = MONTHS[m[4]];
  if (!mon) throw new Error(`Unknown month: ${m[4]}`);
  const now = new Date();
  let cand = new Date(now.getFullYear(), mon - 1, dd, hh, mm, 0, 0);
  if (cand <= now) cand = new Date(now.getFullYear() + 1, mon - 1, dd, hh, mm, 0, 0);
  return Math.floor(cand.getTime() / 1000);
}

function resetStrToEpoch(s) {
  if (!s) return null;
  const trimmed = s.trim();
  if (/^\d{1,2}:\d{2}$/.test(trimmed)) return nextLocalTimeToEpoch(trimmed);
  if (trimmed.includes(" on ")) {
    try {
      return parseNoYearToEpoch(trimmed);
    } catch {
      return null;
    }
  }
  return null;
}

const RE_5H = /5h limit:.*?(\d+)% left(?:.*?resets\s+([0-9]{1,2}:[0-9]{2}))?/i;
const RE_WEEK = /Weekly limit:.*?(\d+)% left/i;
const RE_WEEK_RESET_NEXT = /^\(?resets\s+(.+?)\)?$/i;
const RE_RESET_INLINE = /resets\s+([0-9]{1,2}:[0-9]{2}(?:\s+on\s+\d{1,2}\s+[A-Za-z]{3})?)/i;

function extractLimits(lines, log) {
  const mk = () => ({
    five_hour_left_percent: null,
    five_hour_resets_at: null,
    weekly_left_percent: null,
    weekly_resets_at: null,
  });
  const main = mk();
  const spark = mk();
  let section = "main";

  for (let i = 0; i < lines.length; i += 1) {
    const ln = lines[i];
    if (ln.includes("GPT-5.3-Codex-Spark limit") || ln.includes("Spark limit")) {
      section = "spark";
      continue;
    }

    const target = section === "main" ? main : spark;

    const m5 = ln.match(RE_5H);
    if (m5 && target.five_hour_left_percent == null) {
      target.five_hour_left_percent = Number(m5[1]);
      target.five_hour_resets_at = resetStrToEpoch(m5[2]);
    }

    const mw = ln.match(RE_WEEK);
    if (mw) {
      if (target.weekly_left_percent == null) {
        target.weekly_left_percent = Number(mw[1]);
      }
      if (target.weekly_resets_at == null) {
        const inline = ln.match(RE_RESET_INLINE);
        if (inline) {
          target.weekly_resets_at = resetStrToEpoch(inline[1]);
        } else if (i + 1 < lines.length) {
          const next = lines[i + 1].match(RE_WEEK_RESET_NEXT);
          if (next) target.weekly_resets_at = resetStrToEpoch(next[1]);
        }
      }
    }
  }

  log(`Parsed main=${JSON.stringify(main)}`);
  log(`Parsed spark=${JSON.stringify(spark)}`);

  return { main, spark };
}

function tailForLog(text, maxChars = 400) {
  const cleaned = text.replace(/\r/g, "\\r").replace(/\n/g, "\\n");
  if (cleaned.length <= maxChars) return cleaned;
  return `...${cleaned.slice(-maxChars)}`;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function respondToTerminalQueries(stdin, text, log) {
  if (text.includes("\x1b[6n")) {
    stdin.write("\x1b[1;1R");
    log("Replied to terminal query ESC[6n with ESC[1;1R.");
  }
  if (text.includes("\x1b[c")) {
    stdin.write("\x1b[?1;2c");
    log("Replied to terminal query ESC[c with ESC[?1;2c.");
  }
  if (text.includes("\x1b]10;?")) {
    stdin.write("\x1b]10;rgb:cccc/cccc/cccc\x1b\\");
    log("Replied to terminal color query OSC 10.");
  }
  if (text.includes("\x1b]11;?")) {
    stdin.write("\x1b]11;rgb:0000/0000/0000\x1b\\");
    log("Replied to terminal color query OSC 11.");
  }
}

async function slowType(stdin, text, delayS) {
  for (const ch of text) {
    stdin.write(ch);
    if (delayS > 0) await sleep(delayS * 1000);
  }
}

async function sendLine(stdin, line, delayS) {
  await slowType(stdin, line, delayS);
  stdin.write("\r");
}

async function waitForMarkers(ctx, opts) {
  const {
    markers,
    timeoutS,
    phase,
    idleReadyBytes = 0,
    idleReadyS = 1.0,
  } = opts;

  const deadline = Date.now() + timeoutS * 1000;
  let nextProgress = Date.now() + 2000;

  while (Date.now() < deadline) {
    if (ctx.exitCode !== null) {
      throw new Error(
        `Codex exited during ${phase} with code ${ctx.exitCode}. Captured tail: ${tailForLog(ctx.buf)}`
      );
    }

    if (ctx.bufChanged) {
      ctx.bufChanged = false;
      const normText = normalizeCapture(ctx.buf.slice(-12000)).join("\n");
      if (markers.some((m) => normText.includes(m))) {
        ctx.log(`Matched marker for ${phase}. bytes_seen=${ctx.bytesSeen}`);
        return;
      }
    }

    const now = Date.now();
    if (idleReadyBytes && ctx.bytesSeen >= idleReadyBytes && (now - ctx.lastChunkAt) >= idleReadyS * 1000) {
      ctx.log(`No new output for ${idleReadyS.toFixed(2)}s after ${ctx.bytesSeen} bytes; treating ${phase} as ready.`);
      return;
    }

    if (now >= nextProgress) {
      ctx.log(`Still waiting (${phase})... bytes_seen=${ctx.bytesSeen}, tail=${tailForLog(ctx.buf)}`);
      nextProgress = now + 2000;
    }

    await sleep(50);
  }

  ctx.log(`Timed out waiting for markers (${phase}): ${JSON.stringify(markers)}. Tail:\n${ctx.buf.slice(-2000)}`);
  throw new Error("Timed out waiting for expected output from Codex.");
}

async function runCodexStatus({ codexPath, slowDelay, readyTimeout, statusTimeout, debug }, log) {
  const args = ["--no-alt-screen", "-c", "check_for_update_on_startup=false"];
  log(`Spawning: ${codexPath} ${args.join(" ")}`);

  const child = pty.spawn(codexPath, args, {
    name: process.env.TERM || "xterm-256color",
    cols: 120,
    rows: 40,
    cwd: process.cwd(),
    env: { ...process.env },
  });

  const ctx = {
    log,
    buf: "",
    captureChunks: [],
    bytesSeen: 0,
    lastChunkAt: Date.now(),
    bufChanged: false,
    exitCode: null,
  };

  child.onData((text) => {
    respondToTerminalQueries(child, text, log);
    ctx.buf += text;
    ctx.captureChunks.push(text);
    ctx.bytesSeen += Buffer.byteLength(text, "utf8");
    ctx.lastChunkAt = Date.now();
    ctx.bufChanged = true;
  });

  child.onExit(({ exitCode }) => {
    if (debug) {
      log(`Codex exited: ${exitCode}`);
    }
    ctx.exitCode = exitCode == null ? 0 : exitCode;
  });

  try {
    log("Waiting for Codex UI to be ready...");
    await waitForMarkers(ctx, {
      markers: [
        "OpenAI Codex (v",
        "model:",
        "/model to change",
        "? for shortcuts",
        "Run /status",
        "› Find and fix a bug",
      ],
      timeoutS: readyTimeout,
      phase: "startup",
      idleReadyBytes: 1200,
      idleReadyS: 1.0,
    });
    log("Ready.");

    log("Sending /status (slow)...");
    await sendLine(child, "/status", slowDelay);

    log("Waiting for /status output...");
    await waitForMarkers(ctx, {
      markers: ["5h limit:", "Weekly limit:", "/codex/settings/usage"],
      timeoutS: statusTimeout,
      phase: "status",
    });

    await sleep(600);

    log("Sending /quit...");
    await sendLine(child, "/quit", slowDelay);

    const ended = await Promise.race([
      new Promise((resolve) => {
        const interval = setInterval(() => {
          if (ctx.exitCode !== null) {
            clearInterval(interval);
            resolve(true);
          }
        }, 50);
      }),
      sleep(2000).then(() => false),
    ]);

    if (!ended) {
      log("Codex did not exit promptly; terminating.");
      child.kill();
      const ended2 = await Promise.race([
        new Promise((resolve) => {
          const interval = setInterval(() => {
            if (ctx.exitCode !== null) {
              clearInterval(interval);
              resolve(true);
            }
          }, 50);
        }),
        sleep(2000).then(() => false),
      ]);
      if (!ended2) {
        log("Codex still running; killing.");
        child.kill();
      }
    }

    return ctx.captureChunks.join("");
  } finally {
    // no-op
  }
}

function getTimezoneLabel() {
  const dtf = new Intl.DateTimeFormat("en-US", { timeZoneName: "short" });
  const parts = dtf.formatToParts(new Date());
  const tzPart = parts.find((p) => p.type === "timeZoneName");
  return tzPart ? tzPart.value : "local";
}

async function main() {
  try {
    const args = parseArgs(process.argv);
    const log = makeLogger(args.debug);

    const raw = await runCodexStatus(args, log);
    const lines = normalizeCapture(raw);
    const limits = extractLimits(lines, log);

    const payload = {
      captured_at: Math.floor(Date.now() / 1000),
      timezone: getTimezoneLabel(),
      limits,
    };

    process.stdout.write(`${JSON.stringify(payload, null, 2)}\n`);
  } catch (err) {
    process.stderr.write(`${err instanceof Error ? err.stack || err.message : String(err)}\n`);
    process.exit(1);
  }
}

main();

File: codex_status.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import json
import os
import pty
import re
import select
import subprocess
import sys
import time
from dataclasses import dataclass
from datetime import datetime, time as dtime, timedelta
from typing import Optional, List


# ----------------------------
# Debug logging
# ----------------------------

def make_logger(debug: bool):
    def log(msg: str):
        if not debug:
            return
        ts = datetime.now().astimezone().strftime("%Y-%m-%d %H:%M:%S%z")
        line = f"[{ts}] {msg}\n"
        sys.stderr.write(line)
        sys.stderr.flush()
    return log


# ----------------------------
# PTY subprocess helpers
# ----------------------------

def set_pty_nonblocking(fd: int):
    import fcntl
    flags = fcntl.fcntl(fd, fcntl.F_GETFL)
    fcntl.fcntl(fd, fcntl.F_SETFL, flags | os.O_NONBLOCK)

def set_pty_winsize(fd: int, rows: int = 40, cols: int = 120):
    import fcntl
    import struct
    import termios
    # struct winsize { unsigned short ws_row, ws_col, ws_xpixel, ws_ypixel; }
    winsz = struct.pack("HHHH", rows, cols, 0, 0)
    fcntl.ioctl(fd, termios.TIOCSWINSZ, winsz)

def read_pty_available(master_fd: int, timeout_s: float) -> str:
    """Read whatever is available on the PTY within timeout_s, return decoded text."""
    chunks: List[bytes] = []
    r, _, _ = select.select([master_fd], [], [], timeout_s)
    if not r:
        return ""
    while True:
        try:
            data = os.read(master_fd, 4096)
            if not data:
                break
            chunks.append(data)
            # If more data is immediately available, keep draining; otherwise stop.
            r2, _, _ = select.select([master_fd], [], [], 0)
            if not r2:
                break
        except BlockingIOError:
            break
        except OSError:
            break
    if not chunks:
        return ""
    return b"".join(chunks).decode("utf-8", errors="replace")

def tail_for_log(text: str, max_chars: int = 400) -> str:
    cleaned = text.replace("\r", "\\r").replace("\n", "\\n")
    if len(cleaned) <= max_chars:
        return cleaned
    return f"...{cleaned[-max_chars:]}"

def respond_to_terminal_queries(master_fd: int, text: str, log) -> None:
    # Some Codex UI stacks probe terminal capabilities and can stall if probes are unanswered.
    # Reply with conservative values to keep startup progressing in raw PTY mode.
    if "\x1b[6n" in text:
        os.write(master_fd, b"\x1b[1;1R")  # Cursor position report: row 1, col 1.
        log("Replied to terminal query ESC[6n with ESC[1;1R.")
    if "\x1b[c" in text:
        os.write(master_fd, b"\x1b[?1;2c")  # Primary DA: VT100 with advanced video option.
        log("Replied to terminal query ESC[c with ESC[?1;2c.")
    if "\x1b]10;?" in text:
        os.write(master_fd, b"\x1b]10;rgb:cccc/cccc/cccc\x1b\\")
        log("Replied to terminal color query OSC 10.")
    if "\x1b]11;?" in text:
        os.write(master_fd, b"\x1b]11;rgb:0000/0000/0000\x1b\\")
        log("Replied to terminal color query OSC 11.")

def slow_type(master_fd: int, text: str, delay_s: float):
    for ch in text:
        os.write(master_fd, ch.encode("utf-8"))
        time.sleep(delay_s)

def send_line(master_fd: int, line: str, delay_s: float):
    slow_type(master_fd, line, delay_s)
    os.write(master_fd, b"\r")


# ----------------------------
# Parsing helpers
# ----------------------------

ANSI_RE = re.compile(r"\x1b\[[0-9;]*[A-Za-z]")
BOX_TRANS = str.maketrans({
    "│": " ", "╭": " ", "╮": " ", "╰": " ", "╯": " ",
    "─": " ", "•": " ", "╳": " ",
})
MONTHS = {m: i for i, m in enumerate(
    ["Jan", "Feb", "Mar", "Apr", "May", "Jun",
     "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"], start=1
)}

RE_5H = re.compile(r"5h limit:.*?(\d+)% left(?:.*?resets\s+([0-9]{1,2}:[0-9]{2}))?", re.I)
RE_WEEK = re.compile(r"Weekly limit:.*?(\d+)% left", re.I)
RE_WEEK_RESET_NEXT = re.compile(r"^\(?resets\s+(.+?)\)?$", re.I)
RE_RESET_INLINE = re.compile(
    r"resets\s+([0-9]{1,2}:[0-9]{2}(?:\s+on\s+\d{1,2}\s+[A-Za-z]{3})?)",
    re.I,
)

def normalize_capture(raw: str) -> List[str]:
    raw = ANSI_RE.sub("", raw)
    raw = raw.translate(BOX_TRANS)
    lines = []
    for ln in raw.splitlines():
        ln = re.sub(r"\s+", " ", ln).strip()
        if ln:
            lines.append(ln)
    return lines

def next_local_time_to_epoch(hhmm: str, tz) -> int:
    h, m = map(int, hhmm.split(":"))
    now = datetime.now().astimezone()
    cand = datetime.combine(now.date(), dtime(h, m), tzinfo=tz)
    if cand <= now:
        cand += timedelta(days=1)
    return int(cand.timestamp())

def parse_no_year_to_epoch(s: str, tz) -> int:
    # "03:49 on 28 Feb"
    m = re.match(r"^(\d{1,2}):(\d{2})\s+on\s+(\d{1,2})\s+([A-Za-z]{3})$", s.strip())
    if not m:
        raise ValueError(f"Unrecognized reset format: {s!r}")
    hh, mm, dd, mon = int(m.group(1)), int(m.group(2)), int(m.group(3)), m.group(4)
    mon_i = MONTHS.get(mon)
    if not mon_i:
        raise ValueError(f"Unknown month: {mon!r}")
    now = datetime.now().astimezone()
    yr = now.year
    cand = datetime(yr, mon_i, dd, hh, mm, tzinfo=tz)
    if cand <= now:
        cand = datetime(yr + 1, mon_i, dd, hh, mm, tzinfo=tz)
    return int(cand.timestamp())

def reset_str_to_epoch(s: Optional[str], tz) -> Optional[int]:
    if not s:
        return None
    s = s.strip()
    if re.fullmatch(r"\d{1,2}:\d{2}", s):
        return next_local_time_to_epoch(s, tz)
    if " on " in s:
        try:
            return parse_no_year_to_epoch(s, tz)
        except Exception:
            return None
    return None

@dataclass
class Limits:
    five_hour_left_percent: Optional[int] = None
    five_hour_resets_at: Optional[int] = None
    weekly_left_percent: Optional[int] = None
    weekly_resets_at: Optional[int] = None

def extract_limits(lines: List[str], log) -> dict:
    tz = datetime.now().astimezone().tzinfo  # system tz

    main = Limits()
    spark = Limits()
    section = "main"

    for i, ln in enumerate(lines):
        if "GPT-5.3-Codex-Spark limit" in ln or "Spark limit" in ln:
            section = "spark"
            continue

        target = main if section == "main" else spark

        m5 = RE_5H.search(ln)
        if m5 and target.five_hour_left_percent is None:
            target.five_hour_left_percent = int(m5.group(1))
            target.five_hour_resets_at = reset_str_to_epoch(m5.group(2), tz)

        mw = RE_WEEK.search(ln)
        if mw:
            if target.weekly_left_percent is None:
                target.weekly_left_percent = int(mw.group(1))
            if target.weekly_resets_at is None:
                minline = RE_RESET_INLINE.search(ln)
                if minline:
                    target.weekly_resets_at = reset_str_to_epoch(minline.group(1), tz)
                elif i + 1 < len(lines):
                    mnext = RE_WEEK_RESET_NEXT.match(lines[i + 1])
                    if mnext:
                        target.weekly_resets_at = reset_str_to_epoch(mnext.group(1), tz)

    log(f"Parsed main={main}")
    log(f"Parsed spark={spark}")

    return {
        "main": {
            "five_hour_left_percent": main.five_hour_left_percent,
            "five_hour_resets_at": main.five_hour_resets_at,
            "weekly_left_percent": main.weekly_left_percent,
            "weekly_resets_at": main.weekly_resets_at,
        },
        "spark": {
            "five_hour_left_percent": spark.five_hour_left_percent,
            "five_hour_resets_at": spark.five_hour_resets_at,
            "weekly_left_percent": spark.weekly_left_percent,
            "weekly_resets_at": spark.weekly_resets_at,
        },
    }


# ----------------------------
# Codex control logic
# ----------------------------

READY_MARKERS = [
    "OpenAI Codex (v",
    "model:",
    "/model to change",
    "? for shortcuts",
    "Run /status",
    "› Find and fix a bug",
]
STATUS_MARKERS = [
    "5h limit:",
    "Weekly limit:",
    "/codex/settings/usage",
]

def wait_for_markers(
    master_fd: int,
    markers: List[str],
    timeout_s: float,
    log,
    capture: List[str],
    proc: Optional[subprocess.Popen] = None,
    phase: str = "wait",
    idle_ready_bytes: int = 0,
    idle_ready_s: float = 1.0,
) -> None:
    deadline = time.monotonic() + timeout_s
    buf = ""
    bytes_seen = 0
    next_progress_log = time.monotonic() + 2.0
    last_chunk_at = time.monotonic()
    while time.monotonic() < deadline:
        if proc is not None:
            rc = proc.poll()
            if rc is not None:
                tail = tail_for_log(buf)
                raise RuntimeError(
                    f"Codex exited during {phase} with code {rc}. "
                    f"Captured tail: {tail}"
                )

        chunk = read_pty_available(master_fd, timeout_s=0.2)
        if chunk:
            respond_to_terminal_queries(master_fd, chunk, log)
            buf += chunk
            bytes_seen += len(chunk.encode("utf-8", errors="replace"))
            capture.append(chunk)
            last_chunk_at = time.monotonic()

            # Marker checks against normalized text are more robust than scanning raw
            # PTY bytes that contain heavy cursor/format control sequences.
            norm_text = "\n".join(normalize_capture(buf[-12000:]))
            if any(m in norm_text for m in markers):
                log(f"Matched marker for {phase}. bytes_seen={bytes_seen}")
                return

        now = time.monotonic()
        if idle_ready_bytes and bytes_seen >= idle_ready_bytes and (now - last_chunk_at) >= idle_ready_s:
            log(
                f"No new output for {idle_ready_s:.2f}s after {bytes_seen} bytes; "
                f"treating {phase} as ready."
            )
            return

        if now >= next_progress_log:
            log(
                f"Still waiting ({phase})... bytes_seen={bytes_seen}, "
                f"tail={tail_for_log(buf)}"
            )
            next_progress_log = now + 2.0

    # include a small tail in logs if debug
    tail = buf[-2000:]
    log(f"Timed out waiting for markers ({phase}): {markers}. Tail:\n{tail}")
    raise TimeoutError("Timed out waiting for expected output from Codex.")

def run_codex_status(
    codex_path: str,
    slow_delay: float,
    ready_timeout: float,
    status_timeout: float,
    debug: bool,
    log,
) -> str:
    master_fd, slave_fd = pty.openpty()
    set_pty_nonblocking(master_fd)
    set_pty_winsize(master_fd)
    set_pty_winsize(slave_fd)

    # Keep startup behavior aligned with the proven bash implementation.
    cmd = [codex_path, "--no-alt-screen", "-c", "check_for_update_on_startup=false"]

    log(f"Spawning: {' '.join(cmd)}")

    # Start in a new session so it behaves like a real terminal program
    proc = subprocess.Popen(
        cmd,
        stdin=slave_fd,
        stdout=slave_fd,
        stderr=slave_fd,
        start_new_session=True,
        close_fds=True,
        env=os.environ.copy(),
    )
    os.close(slave_fd)

    capture_chunks: List[str] = []

    try:
        log("Waiting for Codex UI to be ready...")
        wait_for_markers(
            master_fd,
            READY_MARKERS,
            ready_timeout,
            log,
            capture_chunks,
            proc=proc,
            phase="startup",
            idle_ready_bytes=1200,
            idle_ready_s=1.0,
        )
        log("Ready.")

        log("Sending /status (slow)...")
        send_line(master_fd, "/status", delay_s=slow_delay)

        log("Waiting for /status output...")
        wait_for_markers(
            master_fd,
            STATUS_MARKERS,
            status_timeout,
            log,
            capture_chunks,
            proc=proc,
            phase="status",
        )

        # Give it a moment to fully render the status box
        time.sleep(0.6)
        extra = read_pty_available(master_fd, timeout_s=0.2)
        if extra:
            capture_chunks.append(extra)

        log("Sending /quit...")
        send_line(master_fd, "/quit", delay_s=slow_delay)

        # Wait briefly for exit, then terminate if needed
        try:
            proc.wait(timeout=2.0)
        except subprocess.TimeoutExpired:
            log("Codex did not exit promptly; terminating.")
            proc.terminate()
            try:
                proc.wait(timeout=2.0)
            except subprocess.TimeoutExpired:
                log("Codex still running; killing.")
                proc.kill()

        return "".join(capture_chunks)

    finally:
        try:
            os.close(master_fd)
        except OSError:
            pass


def main():
    ap = argparse.ArgumentParser(description="Spawn Codex in a PTY, run /status, print limits JSON to stdout.")
    ap.add_argument("--codex-path", default="codex", help="Path to codex executable.")
    ap.add_argument("--slow-delay", type=float, default=0.10, help="Delay per character when typing commands.")
    ap.add_argument("--ready-timeout", type=float, default=30.0, help="Seconds to wait for Codex UI to appear.")
    ap.add_argument("--status-timeout", type=float, default=30.0, help="Seconds to wait for /status output.")
    ap.add_argument("--debug", action="store_true", help="Enable debug logs.")
    args = ap.parse_args()

    log = make_logger(args.debug)

    raw = run_codex_status(
        codex_path=args.codex_path,
        slow_delay=args.slow_delay,
        ready_timeout=args.ready_timeout,
        status_timeout=args.status_timeout,
        debug=args.debug,
        log=log,
    )

    lines = normalize_capture(raw)
    limits = extract_limits(lines, log)

    tz = datetime.now().astimezone().tzinfo
    payload = {
        "captured_at": int(datetime.now().astimezone().timestamp()),
        "timezone": str(tz),
        "limits": limits,
    }

    print(json.dumps(payload, indent=2))


if __name__ == "__main__":
    main()
